In [206]:
import pandas as pd
import openai
import os
from openai.embeddings_utils import get_embedding, cosine_similarity
import numpy as np


In [207]:
openai.api_key = os.getenv("OPENAI_API_KEY")
openai.api_type = "azure"
openai.api_base = "https://cog-fxpoc-tonality-dev-01.openai.azure.com/"
openai.api_version = "2023-03-15-preview"

In [208]:
product_names = ["Uncrewed Surface Vehicle, Sounder"]

In [209]:
df = pd.read_excel("./data/AI_products.xlsx")

df

Unnamed: 0                                       Product_Name  \
0             0                        AIS physical shore stations   
1             1                      AIS service management system   
2             2                                AIS space receivers   
3             3  APOS - Operating station for HiPAP and HPR Aco...   
4             4  APOS Survey - Surveyor's independent Operator ...   
..          ...                                                ...   
224         224                                              MRU E   
225         225                                              MRU 5   
226         226                                              MRU H   
227         227                                            MRU 5+    
228         228                                              MRU S   

                                      Product category  \
0                       Communication solutions, Naval   
1                            Surveillance & monitoring   
2                            Surveillance & monitoring   
3                            Surveillance & monitoring   
4                              Communication solutions   
..                                                 ...   
224  Geophysical survey, Seafloor mapping, Surveill...   
225    Ocean science, Surface navigation & positioning   
226    Ocean science, Surface navigation & positioning   
227    Ocean science, Surface navigation & positioning   
228    Ocean science, Surface navigation & positioning   

                                          General text  \
0                                                  NaN   
1    Monitoring of all AIS vessels and AIS AtoN wit...   
2                                                  NaN   
3    The product brings significant value to users ...   
4    APOS Survey offers full LBL and SSBL, quicker ...   
..                                                 ...   
224  The MRU E is an extended temperature motion se...   
225  THE IDEAL MARINE MOTION SENSOR. The fifth gene...   
226  THE VERSATILE HEAVE COMPENSATOR This fifth gen...   
227  The MRU 5+ product is upgraded with the best M...   
228  The MRU S is a roll, pitch and heave motion se...   

                                              Features  \
0                                                  NaN   
1     - Monitoring of all AIS vessels and AIS AtoN ...   
2                                                  NaN   
3     - Innovative design\n - Advanced functionalit...   
4    APOS Survey offers: \n- Full LBL and SSBL \n- ...   
..                                                 ...   
224   - 0.05° roll and pitch accuracy\n- Outputs re...   
225  • 0.02° roll and pitch accuracy. • Outputs on ...   
226  • 0.05° roll and pitch accuracy • High accurac...   
227  • 0.008° roll and pitch accuracy\n• Exceptiona...   
228  • 0.3° roll and pitch accuracy \n• Outputs rea...   

                              Technical Specifications  \
0                                                  NaN   
1                                         I don't know   
2                                                  NaN   
3     - Category: Design\n   - Parameters: Innovati...   
4    MP8300 computer: Type-approved by ABS, BV, CCS...   
..                                                 ...   
224  MRU E, ORIENTATION OUTPUT INPUT FORMATS\n- Ang...   
225  MRU 5. ORIENTATION OUTPUT: Analog channels (ju...   
226  MRU H ORIENTATION Angular orientation range ±1...   
227  MRU 5+ MK-II\nORIENTATION OUTPUT Analog channe...   
228  MRU S \nORIENTATION OUTPUT Analog channels (ju...   

                                                   url  \
0    https://www.kongsberg.com/maritime/products/on...   
1    https://www.kongsberg.com/maritime/products/on...   
2    https://www.kongsberg.com/maritime/products/on...   
3    https://www.kongsberg.com/maritime/products/Ac...   
4    https://www.kongsberg.com/maritime/products/Ac...   
..                          

In [210]:
df[df["Product_Name"] == product_names[0]]

Unnamed: 0                       Product_Name  \
169         169  Uncrewed Surface Vehicle, Sounder   

                Product category  \
169  Fish finding, Ocean science   

                                          General text  \
169  The product offers a unique hull design for op...   

                                              Features  \
169  • Unique hull design for optimal performance\n...   

                              Technical Specifications  \
169  Length: 8m\nBeam: 2.2m\nHeight: 2.3/4.4m (mast...   

                                                   url  \
169  https://www.kongsberg.com/maritime/products/ma...   

                                           Data sheets  \
169  https://www.kongsberg.com/contentassets/67e209...   

                                             downloads  
169  {'Data sheet': ['https://www.kongsberg.com/con...

In [211]:
filtered_df = df.loc[[9, 10, 11, 12, 55, 56, 169, 170]]

In [212]:
filtered_df

Unnamed: 0                                    Product_Name  \
9             9            Autonomous Underwater Vehicle, HUGIN   
10           10       Autonomous Underwater Vehicle, HUGIN Edge   
11           11  Autonomous Underwater Vehicle, HUGIN Endurance   
12           12   Autonomous Underwater Vehicle, HUGIN Superior   
55           55                        Kongsberg Camera Cluster   
56           56                  Kongsberg Proximity Assistance   
169         169               Uncrewed Surface Vehicle, Sounder   
170         170             Underwater Altimeters - 1107 series   

                                      Product category  \
9                            Surveillance & monitoring   
10   Seafloor mapping, Surveillance & monitoring, N...   
11   Seafloor mapping, Surveillance & monitoring, N...   
12   Seafloor mapping, Surveillance & monitoring, N...   
55                    Surface navigation & positioning   
56                      Communication solutions, Naval   
169                        Fish finding, Ocean science   
170                        Fish finding, Ocean science   

                                          General text  \
9                                                  NaN   
10   The HUGIN Edge is a next generation mid-size A...   
11   HUGIN Endurance is the newest member of the HU...   
12   Lorem ipsum dolor sit amet, consectetur adipis...   
55   UNCOMPROMISING MARITIME IMAGERY The navigator’...   
56   PROXIMITYVIEW 100 VISUALIZE THE SITUATION The ...   
169  The product offers a unique hull design for op...   
170  The KONGSBERG underwater altimeter product lin...   

                                              Features  \
9                                                  NaN   
10   • Exceptional reliability, durability and perf...   
11   HUGIN Endurance can be configured with a varie...   
12    - Feature 1: Lorem ipsum dolor sit amet, cons...   
55    - Compact and hard-wearing housing\n - Config...   
56   Dynamic stitching of several KONGSBERG KCC. St...   
169  • Unique hull design for optimal performance\n...   
170  Robust design, Easily configurable analog / di...   

                              Technical Specifications  \
9                                                  NaN   
10   Endurance: 24 hours @ 4 knots, 85% payload usa...   
11   HUGIN Endurance is capable of mission ranges u...   
12    - Category: Lorem ipsum dolor sit amet - Para...   
55   PERFORMANCE POWER\n - Image frame rate: 1/2/4 ...   
56   PERFORMANCE ENVIRONMENTAL SPECIFICATIONS - Fus...   
169  Length: 8m\nBeam: 2.2m\nHeight: 2.3/4.4m (mast...   
170                                       I don't know   

                                                   url  \
9    https://www.kongsberg.com/maritime/products/ma...   
10   https://www.kongsberg.com/maritime/products/ma...   
11   https://www.kongsberg.com/maritime/products/ma...   
12   https://www.kongsberg.com/maritime/products/ma...   
55   https://www.kongsberg.com/maritime/products/si...   
56   https://www.kongsberg.com/maritime/products/si...   
169  https://www.kongsberg.com/maritime/products/ma...   
170  https://www.kongsberg.com/maritime/products/ma...   

                                           Data sheets  \
9                                                  NaN   
10   https://www.kongsberg.com/contentassets/8cf870...   
11   https://www.kongsberg.com/contentassets/3a18a1...   
12   https://www.kongsberg.com/globalassets/maritim...   
55   https://www.kongsberg.com/contentassets/e66792...   
56   https://www.kongsberg.com/contentassets/9389e7...   
169  https://www.kongsberg.com/contentassets/67e209...   
170                                                NaN   

                                             downloads  
9    {'Course description': ['https://www.kongsberg...  
10   {'Data sheet': ['https://www.kongsberg.com/con...  
11   {'Data sheet': ['https://www.kongsberg.com/con...  
12   {'Course description': ['https

In [213]:
df_filters = pd.read_excel("filter_tags_12_12.xlsx")

df_filters

category  \
0     Autonomous and uncrewed solutions   
1                        Cloud solution   
2               Communication solutions   
3                          Fish-finding   
4                         Ocean science   
5                    Geophysical survey   
6                    Inertial solutions   
7                                 Naval   
8                      Seafloor mapping   
9      Surface navigation & positioning   
10            Surveillance & monitoring   
11  Underwater navigation & positioning   

                                         product type  \
0   AUV (aut) products, USV(und) products, Hugin A...   
1   Sensor Remote, Data Processing, Data Managemen...   
2                 Modem, Broadband radio, Transceiver   
3   Fish-finding echo sounder, Sonar, Transducers,...   
4   Scientific Echosounders,  Scientific multibeam...   
5   Autonomous Underwater Vehicle, Multibeam echo ...   
6   MRU (Motion Reference Unit), MGC (Motion Gyro ...   
7   Autonomous Underwater Vehicle, Sonar, Uncrewed...   
8   Multibeam echo sounder, Single beam echo sound...   
9   GPS receiver, IALA beacon receiver, Inertial S...   
10  Acoustic Control System, Autonomous Underwater...   
11  Modem, Navigation system, Positioning system, ...   

                                           technology  \
0   Navigation, 3000 meter, 4500 meter, 6000 meter...   
1   Sensor data analysis, Cloud-based, AI classifi...   
2                    Acoustic Modem, VDES technology,   
3   Split-Beam transducer, Single beam transducer,...   
4   Wideband technology, Acoustic Doppler, GPS int...   
5      Positioning & attitude, Multibeam, Single beam   
6   inertial measurement, MEMS gyro, motion sensor...   
7   Customizable Payload, Remote Vessel Operation,...   
8   High-Resolution Sonar, Multibeam  Echo Sounder...   
9   Augmented Reality, GNSS infrastructure, Sensor...   
10  3D sonar technology, Acoustic Telemetry Link, ...   
11  Acoustic Telemetry, Hydroacoustic Positioning,...   

                                          application  
0   Environmental Monitoring, Marine Survey, Marit...  
1   Ocean Data collection, Environmental monitorin...  
2   Maritime Communication, Subsea Communication, ...  
3       Fish Finding, Hydroacoustic Synchronization,   
4   Scientific research, Environmental monitoring,...  
5   Deep-Sea Exploration, Seafloor Mapping, Sedime...  
6   Sensor data optimization, Post-processing anal...  
7   Autonomous Maritime Vehicle, Long Endurance, M...  
8   Deep Sea Exploration, Oceanography, Seafloor M...  
9   Environmental Monitoring, GPS, Maritime Naviga...  
10  BOP operation, Marine Survey, Remote Sensing, ...  
11  Dynamic Positioning Reference, ROV integration...

In [214]:
hierarchy = {}
for category, group in df_filters.groupby('category'):
    hierarchy[category] = {}
    for _, row in group.iterrows():
        prod_type =  row['technology']
    hierarchy[category] = prod_type

In [215]:
hierarchy

{'Autonomous and uncrewed solutions': 'Navigation, 3000 meter, 4500 meter, 6000 meter, Multibeam Echo sounder, Single beam echo sounder, Sub-bottom profiler, Situational awareness',
 'Cloud solution': 'Sensor data analysis, Cloud-based, AI classification, Cyber Security',
 'Communication solutions': 'Acoustic Modem, VDES technology,',
 'Fish-finding': 'Split-Beam transducer, Single beam transducer, Multi-frequency transducer, Fish-finding sonar, Purse seine & trawling sonar, Matrix Sonar, Trawl sonar, Split beam echo sounder, Single beam echo sounder, Gear monitoring sensors',
 'Geophysical survey': 'Positioning & attitude, Multibeam, Single beam',
 'Inertial solutions': 'inertial measurement, MEMS gyro, motion sensor, attitude determination',
 'Naval': 'Customizable Payload, Remote Vessel Operation, USV hydroacoustic applications, Diver Detection, Mine avoidance',
 'Ocean science': 'Wideband technology, Acoustic Doppler, GPS integrated system, Split-beam transducer, Autonomous transce

In [216]:
hierarchy["Autonomous and uncrewed solutions"]

'Navigation, 3000 meter, 4500 meter, 6000 meter, Multibeam Echo sounder, Single beam echo sounder, Sub-bottom profiler, Situational awareness'

In [217]:
df = filtered_df.copy()

In [218]:
df

Unnamed: 0                                    Product_Name  \
9             9            Autonomous Underwater Vehicle, HUGIN   
10           10       Autonomous Underwater Vehicle, HUGIN Edge   
11           11  Autonomous Underwater Vehicle, HUGIN Endurance   
12           12   Autonomous Underwater Vehicle, HUGIN Superior   
55           55                        Kongsberg Camera Cluster   
56           56                  Kongsberg Proximity Assistance   
169         169               Uncrewed Surface Vehicle, Sounder   
170         170             Underwater Altimeters - 1107 series   

                                      Product category  \
9                            Surveillance & monitoring   
10   Seafloor mapping, Surveillance & monitoring, N...   
11   Seafloor mapping, Surveillance & monitoring, N...   
12   Seafloor mapping, Surveillance & monitoring, N...   
55                    Surface navigation & positioning   
56                      Communication solutions, Naval   
169                        Fish finding, Ocean science   
170                        Fish finding, Ocean science   

                                          General text  \
9                                                  NaN   
10   The HUGIN Edge is a next generation mid-size A...   
11   HUGIN Endurance is the newest member of the HU...   
12   Lorem ipsum dolor sit amet, consectetur adipis...   
55   UNCOMPROMISING MARITIME IMAGERY The navigator’...   
56   PROXIMITYVIEW 100 VISUALIZE THE SITUATION The ...   
169  The product offers a unique hull design for op...   
170  The KONGSBERG underwater altimeter product lin...   

                                              Features  \
9                                                  NaN   
10   • Exceptional reliability, durability and perf...   
11   HUGIN Endurance can be configured with a varie...   
12    - Feature 1: Lorem ipsum dolor sit amet, cons...   
55    - Compact and hard-wearing housing\n - Config...   
56   Dynamic stitching of several KONGSBERG KCC. St...   
169  • Unique hull design for optimal performance\n...   
170  Robust design, Easily configurable analog / di...   

                              Technical Specifications  \
9                                                  NaN   
10   Endurance: 24 hours @ 4 knots, 85% payload usa...   
11   HUGIN Endurance is capable of mission ranges u...   
12    - Category: Lorem ipsum dolor sit amet - Para...   
55   PERFORMANCE POWER\n - Image frame rate: 1/2/4 ...   
56   PERFORMANCE ENVIRONMENTAL SPECIFICATIONS - Fus...   
169  Length: 8m\nBeam: 2.2m\nHeight: 2.3/4.4m (mast...   
170                                       I don't know   

                                                   url  \
9    https://www.kongsberg.com/maritime/products/ma...   
10   https://www.kongsberg.com/maritime/products/ma...   
11   https://www.kongsberg.com/maritime/products/ma...   
12   https://www.kongsberg.com/maritime/products/ma...   
55   https://www.kongsberg.com/maritime/products/si...   
56   https://www.kongsberg.com/maritime/products/si...   
169  https://www.kongsberg.com/maritime/products/ma...   
170  https://www.kongsberg.com/maritime/products/ma...   

                                           Data sheets  \
9                                                  NaN   
10   https://www.kongsberg.com/contentassets/8cf870...   
11   https://www.kongsberg.com/contentassets/3a18a1...   
12   https://www.kongsberg.com/globalassets/maritim...   
55   https://www.kongsberg.com/contentassets/e66792...   
56   https://www.kongsberg.com/contentassets/9389e7...   
169  https://www.kongsberg.com/contentassets/67e209...   
170                                                NaN   

                                             downloads  
9    {'Course description': ['https://www.kongsberg...  
10   {'Data sheet': ['https://www.kongsberg.com/con...  
11   {'Data sheet': ['https://www.kongsberg.com/con...  
12   {'Course description': ['https

In [219]:
df["Product category"]

9                              Surveillance & monitoring
10     Seafloor mapping, Surveillance & monitoring, N...
11     Seafloor mapping, Surveillance & monitoring, N...
12     Seafloor mapping, Surveillance & monitoring, N...
55                      Surface navigation & positioning
56                        Communication solutions, Naval
169                          Fish finding, Ocean science
170                          Fish finding, Ocean science
Name: Product category, dtype: object

In [220]:
text = df["General text"].iloc[4] + df["Features"].iloc[4] + df["Technical Specifications"].iloc[4]

In [221]:
text

'UNCOMPROMISING MARITIME IMAGERY The navigator’s eyes are always on watch. To replace or assist them a camera platform with uncompromising precision and reliability is needed. KONGSBERG Camera Cluster (KCC) delivers just this, in a compact and hard-wearing housing, with configurable 107 to 215 degrees coverage. We utilize the latest in machine vision CMOS technology, together with powerful edge processing, to enable crystal clear imagery. KCC Compliance to EMCD, IACS E10 standard used IEC61010-1:2010+A1:2016. - Compact and hard-wearing housing\n - Configurable 107 to 215 degrees coverage\n - Latest machine vision CMOS technology\n - Powerful edge processingPERFORMANCE POWER\n - Image frame rate: 1/2/4 FPS\n - Supply voltage: 12-30 VDC\n - Mean power consumption: 35 W\n\nENVIRONMENTAL SPECIFICATIONS\n - Operational temperature range: -20°C - +55°C\n - Storage temperature range: -30°C - +80°C\n - Enclosure protection: IP56\n - Gigabit Ethernet\n\nWEIGHTS AND DIMENSIONS\n - L x W x H: 237

In [222]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

filters = hierarchy["Autonomous and uncrewed solutions"].split(",") + hierarchy["Surface navigation & positioning"].split(",") + hierarchy["Surveillance & monitoring"].split(",")

# Combining the text and categories for TF-IDF vectorization
combined_texts = [text] + filters

# Creating the TF-IDF Vectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(combined_texts)

# Calculating cosine similarity between the text and each category
cosine_similarities = cosine_similarity(tfidf_matrix[0:1], tfidf_matrix[1:]).flatten()

# Mapping each category with its similarity score
category_similarity = dict(zip(filters, cosine_similarities))

# Sorting categories based on similarity score
sorted_category_similarity = sorted(category_similarity.items(), key=lambda item: item[1], reverse=True)

sorted_category_similarity



[('3D sonar technology', 0.055649109684259485),
 (' Sensor Fusion Compatible', 0.027824554842129742),
 ('Navigation', 0.0),
 (' 3000 meter', 0.0),
 (' 4500 meter', 0.0),
 (' 6000 meter', 0.0),
 (' Multibeam Echo sounder', 0.0),
 (' Single beam echo sounder', 0.0),
 (' Sub-bottom profiler', 0.0),
 (' Situational awareness', 0.0),
 ('Augmented Reality', 0.0),
 (' GNSS infrastructure', 0.0),
 (' Spoofing Mitigation', 0.0),
 (' Acoustic Telemetry Link', 0.0),
 (' AIS transponder', 0.0),
 (' Autonomous Navigation', 0.0)]

UNCOMPROMISING MARITIME IMAGERY The navigator’s eyes are always on watch. To replace or assist them a camera platform with uncompromising precision and reliability is needed. KONGSBERG Camera Cluster (KCC) delivers just this, in a compact and hard-wearing housing, with configurable 107 to 215 degrees coverage. We utilize the latest in machine vision CMOS technology, together with powerful edge processing, to enable crystal clear imagery. KCC Compliance to EMCD, IACS E10 standard used IEC61010-1:2010+A1:2016.

 - Compact and hard-wearing housing
 - Configurable 107 to 215 degrees coverage
 - Latest machine vision CMOS technology
 - Powerful edge processing

 PERFORMANCE POWER
 - Image frame rate: 1/2/4 FPS
 - Supply voltage: 12-30 VDC
 - Mean power consumption: 35 W

ENVIRONMENTAL SPECIFICATIONS
 - Operational temperature range: -20°C - +55°C
 - Storage temperature range: -30°C - +80°C
 - Enclosure protection: IP56
 - Gigabit Ethernet

WEIGHTS AND DIMENSIONS
 - L x W x H: 237 x 322 x 305 mm (Camera Cluster)
 - Sensor weight: 1.98 kg
 - Mounting bracket weight: 6.98 kg
 - L x W x H: 168 x 268 x 100 mm (Junction Box)
 - Weight: 3.2 kg (incl. 230VAC Power supply)